### PR TITLE
Refactored codebase to address security issues and to help with future change requests

### PR DESCRIPTION
**Pull request by tmancey**
fixes https://github.com/brave-intl/bat-native-confirmations/issues/35
fixes https://github.com/brave-intl/bat-native-confirmations/issues/37
fixes https://github.com/brave-intl/bat-native-confirmations/issues/51

There are a few inconsistencies with passing base64 and/or raw challenge bypass FFI tokens, I will work with eV on best way to resolve this but for now this will not cause us any issues

### CODE REVIEW

File: src/payout_tokens.cc [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  std::vector<std::string> headers = {};
+
+  auto accept_header = std::string("accept: ").append("application/json");
+  headers.push_back(accept_header);
+
+  // Create request content type
+  std::string content_type = "application/json";
+
+  // PUT /v1/confirmation/payment/{payment_id}
+  BLOG(INFO) << "PUT /v1/confirmation/payment/{payment_id}";
```
Comment (emerick) [inline src/payout_tokens.cc]: Is this intentional or did we want to log the actual payment ID here?

File: src/payout_tokens.cc [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  BLOG(INFO) << "URL Request Response:";
+  BLOG(INFO) << "  URL: " << url;
+  BLOG(INFO) << "  Response Status Code: " << response_status_code;
+  BLOG(INFO) << "  Response: " << response;
+  BLOG(INFO) << "  Headers:";
+  for (const auto& header : headers) {
+    BLOG(INFO) << "    " << header.first << ": " << header.second;
+  }
+
+  if (response_status_code != 200) {
```
Comment (emerick) [inline src/payout_tokens.cc]: Are any other 2XX response codes possible here?

File: src/payout_tokens.cc [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+      std::move(payment_credentials_list));
+
+  body_dictionary.SetKey("payload", base::Value(payload_json));
+
+  std::string json;
+  base::JSONWriter::Write(body_dictionary, &json);
+  std::string body = json;
+
+  // Create request headers
+  std::vector<std::string> headers = {};
```
Comment (emerick) [inline src/payout_tokens.cc]: Does explicitly initializing this serve any purpose?

File: src/security_helper.cc [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+#include "tweetnacl.h"
+#include <openssl/base64.h>
+#include <openssl/digest.h>
+#include <openssl/hkdf.h>
+#include <openssl/sha.h>
+
+namespace helper {
+
+std::string Security::Sign(
+    const std::vector<std::string> keys,
```
Comment (emerick) [inline src/security_helper.cc]: I think you want these vectors (keys and values) to be const references.

File: src/confirmations_impl.h [hunk 5]
```diff
@@ -7,218 +7,116 @@
...
 
   std::string ToJSON();
-  bool FromJSON(std::string json);
+  bool FromJSON(const std::string& json);
 
-  void VectorConcat(
-      std::vector<std::string>* dest,
-      std::vector<std::string>* source);
-  std::unique_ptr<base::ListValue> Munge(std::vector<std::string> v);
+  std::unique_ptr<base::ListValue> Munge(const std::vector<std::string>& v);
```
Comment (emerick) [inline src/confirmations_impl.h]: Haha, does it need a more descriptive name? :-)

File: src/redeem_token.cc [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+  std::vector<std::string> headers = {};
+
+  auto accept_header = std::string("accept: ").append("application/json");
+  headers.push_back(accept_header);
+
+  // Create request content type
+  std::string content_type = "application/json";
+
+  // POST /v1/confirmation/{confirmation_id_}/{credential}
+  BLOG(INFO) << "POST /v1/confirmation/{confirmation_id_}/{credential}";
```
Comment (emerick) [inline src/redeem_token.cc]: Same comment as before, jsut checking if this is intentional (not printing the variables).

File: src/refill_tokens.cc [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+namespace confirmations {
+
+RefillTokens::RefillTokens(
+    ConfirmationsImpl* confirmations,
+    ConfirmationsClient* confirmations_client) :
+    public_key_(nullptr),
+    confirmations_(confirmations),
+    confirmations_client_(confirmations_client) {
+  BLOG(INFO) << "Initializing refill tokens";
+  BLOG(INFO) << "Initialized refill tokens";
```
Comment (emerick) [inline src/refill_tokens.cc]: Duplicate log entry here.

File: src/payout_tokens.cc (hunk 1 above)
Comment (tmancey) [reply src/payout_tokens.cc]: This is intentional as the values are logged as part of the URL, thanks

File: src/payout_tokens.cc (hunk 2 above)
Comment (tmancey) [reply src/payout_tokens.cc]: Not according to original code, and checking with sever team on a previous occasion

File: src/payout_tokens.cc (hunk 1 above)
Comment (bridiver) [reply src/payout_tokens.cc]: should be flagged for privacy review?

File: src/confirmations_impl.h (hunk 5 above)
Comment (tmancey) [reply src/confirmations_impl.h]: This is one part of the code I did not refactor much but do agree at some point this should be improved

File: src/payout_tokens.cc (hunk 3 above)
Comment (tmancey) [reply src/payout_tokens.cc]: a habit, all fixes throughout codebase 👍

File: src/refill_tokens.cc (hunk 7 above)
Comment (tmancey) [reply src/refill_tokens.cc]: One is initializing then a function is called, then initialized just for clarity

File: src/security_helper.cc (hunk 4 above)
Comment (tmancey) [reply src/security_helper.cc]: Not possible to my knowledge we can have vector<const T> until C++17?

File: src/payout_tokens.cc (hunk 1 above)
Comment (tmancey) [reply src/payout_tokens.cc]: @bridiver I should have mentioned confirmations is currently undergoing privacy/security review

File: src/payout_tokens.cc (hunk 2 above)
Comment (tmancey) [reply src/payout_tokens.cc]: I have checked and we are covering all response status codes

File: src/security_helper.cc (hunk 4 above)
Comment (emerick) [reply src/security_helper.cc]: I'm just suggesting `const std::vector<std::string>& keys` rather than `const std::vector<std::string> keys`.

Comment (tmancey) [reply src/security_helper.cc]: 👍 yeap missed this one, thanks

**Pull request merged**